designs/initial.md

```diff
@@ -0,0 +1,60 @@
+## `@rest` directive
+
+### Arguments
+
+- route: path to rest endpoint. This could be a path or a full url. If a path, add to the endpoint given on link creation or from the context
+- params: a map of variables to url params
+- method: the HTTP method to send the request via (i.e GET, PUT, POST)
+- type: The GraphQL type this will return
+
+### Notes
+
+It's important that the rest directive could be used at any depth in a query, but once it is used, nothing nested in it can be graphql data, it has to be from the rest link or other resource (like a @client directive)
+
+## `@export` directive
+
+### Arguments
+- as: the string name to create this as a variable to be used down the selection set
+
+### Notes
+
+These are the same semantics that will be supported on the server, but when used in a rest link you can use the exported variables for futher calls (i.e. waterfall requests from nested fields)
+
+```js
+const QUERY = gql`
+  query RestData($email: String!) {
+    users @rest(route: '/users/email/:email', params: { email: $email }, method: 'GET', type: 'User') {
+      id @export(as: "id")
+      firstName
+      lastName
+      friends @rest(route: '/friends/:id', params: { id: $id }, type: '[User]') {
+        firstName
+        lastName
+      }
+    }
+  }
+`;
+```
+
+
+## `createRestLink`
+
+### Arguments
+
+- fetch: an optional implementation of `fetch` (see the http-link for api / warnings). Will use global if found
+- fieldNameNormalizer: a function that takes the response field name and turns into a GraphQL compliant name,for instance "MyFieldName:IsGreat" => myFieldNameIsGreat
+- endpoint: a root endpoint to apply routes to: i.e. api.example.com/v1
```

Comment:
Can we make endpoints a map with a default option so users can add multiple endpoints to their link? If we go this route, we should also add the endpoint key as an argument to the directive.

Reply:
This is a brilliant idea, and not that difficult to implement